pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java

```diff
@@ -0,0 +1,249 @@
+    List<RexNode> operands = new ArrayList<>(function.getOperands());
+    int numOperands = operands.size();
+    for (int i = 0; i < numOperands; i++) {
+      // The RexCall is guaranteed to have all operands as RexLiteral or evaluated as RexLiteral.
+      // So recursively call evaluateLiteralOnlyFunction on all operands.
+      RexNode operand = evaluateLiteralOnlyFunction(rexBuilder, operands.get(i), evaluatedResults);
+      operands.set(i, operand);
+    }
+
+    String functionName = PinotRuleUtils.extractFunctionName(function);
+    FunctionInfo functionInfo = FunctionRegistry.getFunctionInfo(functionName, numOperands);
+    RexNode resultRexNode = rexNode;
+    if (functionInfo != null) {
+      Object[] arguments = new Object[numOperands];
+      for (int i = 0; i < numOperands; i++) {
+        RexNode operand = function.getOperands().get(i);
+        Preconditions.checkArgument(operand instanceof RexLiteral, "Expected all the operands to be RexLiteral");
+        Object value = ((RexLiteral) operand).getValue();
+        if (value instanceof NlsString) {
+          arguments[i] = ((NlsString) value).getValue();
+        } else if (value instanceof GregorianCalendar) {
+          arguments[i] = ((GregorianCalendar) value).getTimeInMillis();
+        } else if (value instanceof ByteString) {
+          arguments[i] = ((ByteString) value).getBytes();
+        } else {
+          arguments[i] = value;
+        }
+      }
+      RelDataType rexNodeType = rexNode.getType();
+      Object functionResult;
+      try {
+        FunctionInvoker invoker = new FunctionInvoker(functionInfo);
+        invoker.convertTypes(arguments);
+        functionResult = invoker.invoke(arguments);
+      } catch (Exception e) {
+        throw new SqlCompilationException(
+            "Caught exception while invoking method: " + functionInfo.getMethod() + " with arguments: "
+                + Arrays.toString(arguments), e);
+      }
+      if (functionResult == null) {
+        resultRexNode = rexBuilder.makeNullLiteral(rexNodeType);
+      } else if (rexNodeType.getSqlTypeName() == SqlTypeName.TIMESTAMP) {
+        long millis = (functionResult instanceof Number) ? ((Number) functionResult).longValue()
+            : TimestampUtils.toMillisSinceEpoch(functionResult.toString());
+        resultRexNode =
+            rexBuilder.makeTimestampLiteral(TimestampString.fromMillisSinceEpoch(millis), rexNodeType.getPrecision());
+      } else if (functionResult instanceof Byte || functionResult instanceof Short || functionResult instanceof Integer
+          || functionResult instanceof Long) {
+        resultRexNode =
+            rexBuilder.makeExactLiteral(BigDecimal.valueOf(((Number) functionResult).longValue()), rexNodeType);
+      } else if (functionResult instanceof Float || functionResult instanceof Double) {
+        resultRexNode =
+            rexBuilder.makeExactLiteral(BigDecimal.valueOf(((Number) functionResult).doubleValue()), rexNodeType);
```

Comment:
This is incorrect. We cannot always convert to double. @zhtaoxiang is fixing this issue in other places
```suggestion
        resultRexNode =
            rexBuilder.makeExactLiteral(new BigDecimal(functionResult.toString()), rexNodeType);
```